README.md
# textlocal

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

## Install

Via Composer

``` bash
$ composer require netsplit/textlocal
```

## Usage

```php
$textlocal = new Netsplit\Textlocal(
    'https://path.to.textlocal.com/api2',
    'api-key-here'
);
```

### Send an SMS message

[API documentation for the "send" endpoint](http://api.txtlocal.com/docs/sendsms).

```php
$response = $textlocal->sendSMS('Message content', [
    // Array of mobile numbers
    '07777777777',
    '07777777778',
], [
    // Optional values
    'containsTrackingLinks' => false, // Should URLs in the SMS be minified?
    'custom'                => 'Any string', // For use as a reference
    'groupID'               => 5, // A Textlocal group ID
    'receiptURL'            => 'http://[...]', // Custom receipt URL
    'scheduleAt'            => 1530735900, // Unix timestamp - send the SMS at this time
    'sender'                => 'Optional Sender Name',
    'sendToOptOut'          => false, // Should the SMS be sent to users who have opted out?
    'simpleReply'           => false, // Disregard sender
    'test'                  => true, // Don't really send the SMS
    'unicode'               => false, // Does the SMS contain unicode characters?
    'validUntil'            => 1530735994, // Unix timestamp - disregard the SMS at this time
]);

// Access values from the response:
echo $response->getStatus()->__toString(); // Hopefully "success"!
echo $response->getBalance()->getValue();
```

### Shorten a URL

[API documentation for the "short URL" endpoint](http://api.txtlocal.com/docs/shorturl)

```php
$response = $textlocal->getShortURL('https://www.google.com');
$shortURL = $response->getShortURL()->__toString();
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email mike@netsplit.org.uk instead of using the issue tracker.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/netsplit/textlocal.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/netsplit/textlocal/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/netsplit/textlocal.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/netsplit/textlocal.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/netsplit/textlocal.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/netsplit/textlocal
[link-travis]: https://travis-ci.org/netsplit/textlocal
[link-scrutinizer]: https://scrutinizer-ci.com/g/netsplit/textlocal/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/netsplit/textlocal
[link-downloads]: https://packagist.org/packages/netsplit/textlocal
[link-author]: https://git.netsplit.uk/mike


CHANGELOG.md
# Changelog

All notable changes to `textlocal` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## [1.1.1] 2018-07-05

### Added
- Renamed support links and moved to a different Gitlab group.

## [1.1.0] 2018-07-04

### Added
- Extra optional arguments for sending messages via SMS.
- Unit tests for Factory classes.

## [1.0.0] 2018-07-04

### Added
- First version.
- The library can now send messages via SMS.


src/Textlocal/Entity/SMS.php
<?php

namespace Netsplit\Textlocal\Textlocal\Entity;

use Netsplit\Textlocal\Textlocal\ValueObject\ContainsTrackingLinks;
use Netsplit\Textlocal\Textlocal\ValueObject\SendToOptOut;
use Netsplit\Textlocal\Textlocal\ValueObject\SimpleReply;
use Netsplit\Textlocal\Textlocal\ValueObject\Test;
use Netsplit\Textlocal\Textlocal\ValueObject\Unicode;
use Netsplit\Textlocal\Textlocal\ValueObject\Custom;
use Netsplit\Textlocal\Textlocal\ValueObject\GroupID;
use Netsplit\Textlocal\Textlocal\ValueObject\ReceiptURL;
use Netsplit\Textlocal\Textlocal\ValueObject\RecipientList;
use Netsplit\Textlocal\Textlocal\ValueObject\ScheduledAt;
use Netsplit\Textlocal\Textlocal\ValueObject\Sender;
use Netsplit\Textlocal\Textlocal\ValueObject\SMSContent;
use Netsplit\Textlocal\Textlocal\ValueObject\ValidUntil;

/**
 * Class SMS describes a message to be sent to a number of mobile numbers.
 *
 * @package Netsplit\Textlocal\Textlocal\Entity
 * @author Mike Jones <mike@netsplit.org.uk>
 * @since 2018-07-04
 */
final class SMS
{
    /**
     * @var array
     */
    public static $optionalOptionMap = [
        'containsTrackingLinks' => 'tracking_links',
        'custom'                => 'custom',
        'groupID'               => 'group_id',
        'receiptURL'            => 'receipt_url',
        'scheduleTime'          => 'schedule_time',
        'sendToOptOut'          => 'optouts',
        'sender'                => 'sender',
        'simpleReply'           => 'simple_reply',
        'test'                  => 'test',
        'unicode'               => 'unicode',
        'validUntil'            => 'validity',
    ];

    /**
     * @var string
     */
    protected $content, $sender, $receiptURL, $custom;

    /**
     * A comma-separated list of mobile numbers.
     *
     * @var string
     */
    protected $recipients;

    /**
     * @var int
     */
    protected $groupID;

    /**
     * UNIX timestamp for a scheduled message.
     *
     * @var int
     */
    protected $scheduleTime;

    /**
     * Overrides the sender with a Simple Reply Service number.
     *
     * @var bool
     */
    protected $simpleReply = false;

    /**
     * UNIX timestamp set to some time in the future. How long a message is
     * valid for.
     *
     * @var int
     */
    protected $validUntil;

    /**
     * Should the message be sent to users who have opted out of your mailing
     * list? Defaults to false.
     *
     * @var bool
     */
    protected $sendToOptOut = false;

    /**
     * Does the message contain unicode characters? Defaults to false.
     *
     * @var bool
     */
    protected $unicode = false;

    /**
     * Does the message contain links which should be converted to short URLs?
     *
     * @var bool
     */
    protected $containsTrackingLinks = false;

    /**
     * If this is set to true, no credit will be used (but the message won't
     * really be sent).
     *
     * @var bool
     */
    protected $test = false;

    /**
     * @param Sender $sender
     * @return void
     */
    public function setSender(Sender $sender)
    {
        $this->sender = $sender->__toString();
    }

    /**
     * @param SMSContent $content
     * @return void
     */
    public function setContent(SMSContent $content)
    {
        $this->content = $content->__toString();
    }

    /**
     * @param RecipientList $list
     * @return void
     */
    public function setRecipientList(RecipientList $list)
    {
        $this->recipients = $list->__toString();
    }

    /**
     * @param SimpleReply $simpleReply
     * @return void
     */
    public function setSimpleReply(SimpleReply $simpleReply)
    {
        $this->simpleReply = $simpleReply->getValue();
    }

    /**
     * @param Test $test
     * @return void
     */
    public function setTest(Test $test)
    {
        $this->test = $test->getValue();
    }

    /**
     * @param Unicode $unicode
     * @return void
     */
    public function setUnicode(Unicode $unicode)
    {
        $this->unicode = $unicode->getValue();
    }

    /**
     * @param ContainsTrackingLinks $containsTrackingLinks
     * @return void
     */
    public function setContainsTrackingLinks(ContainsTrackingLinks $containsTrackingLinks)
    {
        $this->containsTrackingLinks = $containsTrackingLinks->getValue();
    }

    /**
     * @param SendToOptOut $sendToOptOut
     * @return void
     */
    public function setSendToOptOut(SendToOptOut $sendToOptOut)
    {
        $this->sendToOptOut = $sendToOptOut->getValue();
    }

    /**
     * @param ScheduledAt $scheduleTime
     * @return void
     */
    public function setScheduleTime(ScheduledAt $scheduleTime)
    {
        $this->scheduleTime = $scheduleTime->getValue();
    }

    /**
     * @param ValidUntil $validUntil
     * @return void
     */
    public function setValidUntil(ValidUntil $validUntil)
    {
        $this->validUntil = $validUntil->getValue();
    }

    /**
     * @param ReceiptURL $receiptURL
     * @return void
     */
    public function setReceiptURL(ReceiptURL $receiptURL)
    {
        $this->receiptURL = $receiptURL->__toString();
    }

    /**
     * @param GroupID $groupID
     * @return void
     */
    public function setGroupID(GroupID $groupID)
    {
        $this->groupID = $groupID->getValue();
    }

    /**
     * @param Custom $custom
     * @return void
     */
    public function setCustom(Custom $custom)
    {
        $this->custom = $custom->__toString();
    }

    /**
     * Translate our class attributes to an array for Textlocal's "send"
     * endpoint.
     *
     * @return array
     */
    public function toServiceArguments()
    {
        $args = [
            'message' => $this->content,
            'numbers' => $this->recipients,
            'sender'  => $this->sender,
        ];

        foreach (self::$optionalOptionMap as $ours => $theirs) {
            $val = $this->$ours;

            if ($val) {
                $args[$theirs] = $val;
            }
        }

        if ($this->simpleReply) {
            unset($args['sender']);
        }

        return $args;
    }
}